### DISCUSSION TITLE
Simply hooking the IRQ.

### DISCUSSION THREAD

**Discussion by Whammo** [jkotlinski/durexforth#443] [Show and tell]
I was surprised how easy this is.
 End your code with $ea31 jmp,
 Your code holds off the IRQ routine until your code is finished.
```forth
: irqpeek $dc0e c@ ;
: irqpoke $dc0e c! ;
: irqoff irqpeek $fe and irqpoke ;
: irqon irqpeek 1 or irqpoke ;

: irqsub ( addr -- )
irqoff $314 ! irqon ; \ irq vector

code flashborder  \ arbitrary asm code
$d020 inc,
$d020 dec,
$ea31 jmp, \ irq vector default 
;code

: irqtest ['] flashborder irqsub ; 
' flashborder irqsub             
: irqrestore $ea31 irqsub ; 

```
 Fixed!

**Comment by Whammo** [comment]
Accordingly, this IRQ routine fires after the default vector, rather than before,
by ~loading~ pushing another state on the stack over the one created by the kernal,
the IRQ returns to your routine, then your routine returns to forth.
Note: V uses SYS to place the cursor , so V will crash durexForth while this is enabled. 

```forth
: irqpeek $dc0e c@ ;
: irqpoke $dc0e c! ;
: irqoff irqpeek $fe and irqpoke ;
: irqon irqpeek 1 or irqpoke ;

: irqsub ( addr -- )
irqoff $314 ! irqon ;

code pstate   \ uses SYS register locations
$15 lda, pha, $14 lda, pha,
sr lda, pha, ar lda, pha,
xr lda, pha, yr lda, pha,
$ea31 jmp,
;code

code flashborder
$d020 inc,
$d020 dec,
$ea81 jmp, \ into restore registers, rti
;code
: irqtest
['] flashborder $14 ! \ set SYS address
['] pstate irqsub
;

```

**Reply by lonetech** [reply]
Can't claim I understand it fully, but the `irqtest` routine works if using compile time `[']`:

    : irqtest ['] flashborder irqsub ;

I think what effectively happens with just `'` is that you get a word that looks at the source where irqtest is called for the word to look up, before calling flashborder directly, which causes one flash followed by severe confusion as the $ea31 routine tries to return from an interrupt.

**Reply by Whammo** [reply]
Thanks, Yes. I thought it was some sort of POSTPONE required. This is definitely worthy of understanding.

**Reply by lonetech** [reply]
Okay, here goes an attempt at detailed understanding. `[']` is documented under Compiling Forth Words, page 20 of the [durexForth 4.0.0](https://github.com/jkotlinski/durexforth/releases/tag/v4.0.0) manual. `'` is documented under Vectored Execution, page 23. They have identical stack effects and both consume a name from the input. I think `['] foo` is shorthand for `[ ' foo ] literal`. The primary difference is thus that `[']` performs part of its task during the compilation. 

If we look them up in the standard we can get some more detail: [`'`](https://forth-standard.org/standard/core/Tick) and [`[']`](https://forth-standard.org/standard/core/BracketTick) (along with a nice pronunciation suggestion, tick or bracket-tick). We see that `[']` splits its job into compilation and run-time parts; the name is read during compilation, and the result is put on the stack during execution. 

So here's an example of reading a word from the input buffer, executing it, and printing its result:

    : rep ' execute . ;
    ok
    rep 1
    1 ok

Note that this doesn't take arbitrary literals - this worked because `1` is a built in word. Also, the name had to occur on the same line. And the input buffer will be used even if we're working indirectly:

    : nestrep 2 rep . ;
    ok
    3 nestrep 1 .
    1 2 3 ok

[Postpone](https://forth-standard.org/standard/core/POSTPONE) does a bit of the opposite of the `[ foo ] literal` recipe. It compiles a word that would otherwise be [immediate](https://forth-standard.org/standard/core/IMMEDIATE)ly evaluated during compile time. It's designed for use with words that will later be used during compilation. Here's an example showing a poor use:

    hex  : pp' postpone ['] ;
    ok
    here . ' 0 . pp' 0 here .
    3b81 832 3b86 ok
    .s
    ok
    3b81 dump
    3b81 20 f3 11 32 08 ...

We see that it has stored 5 bytes at `here`, which is where words normally get compiled to. The particular values match `' 0 literal`:

    ' 0 literal
    ok
    3b81 dump
    3b81 20 f3 11 32 08 20 f3 11 ...
    3b89 32 08 ...

It's the machine code form of:

    jsr LIT
    !word ZERO

`lit` is a subroutine that reads a word from the call site and puts it on the stack, effectively the 16-bit Forth load immediate. Thus `[']` didn't have stack effects but generated the compiled code that would produce the effect at run-time. 

Since we weren't compiling a word, this was rather inappropriate use of memory. Words like `literal`, `then` and `[']` are only intended for use during compilation, and words like `lit` are only intended for run-time from a compiled word. `postpone`, `immediate`, `[` and `]` manipulate whether words are executed or compiled inside of compilation. 

This flexibility allows words inside Forth to extend the language quite substantially. Here's the actual source code of `[']` from base.fs:

    : ['] ' [ ' literal compile, ]
    ; immediate

As far as I can tell, this could be simpler expressed as:

    : ['] ' postpone literal ; immediate

But it isn't defined that way because `postpone` uses `[']`, is defined later, and probably compiles slightly slower.

**Reply by Whammo** [reply]
Thank you for the generous observations. 
   This requires some rumination.

**Reply by Whammo** [reply]
Also illuminating:
```asm
    +BACKLINK "'", 1
    jsr PARSE_NAME
    jsr TWODUP
    jsr FIND_NAME
    inx
    lda MSB-1,x
    beq print_word_not_found_error
; misplaced assembler directive?
+   ldy LSB-1, x ; in interpreter.asm
; no associated branch instruction? 
    sty LSB, x
    sta MSB, x
    sty LSB+1, x
    sta MSB+1, x
    jsr TO_XT
    jsr SWAP
    jsr GET_IMMED
    inx
    rts
```

**Reply by Whammo** [reply]
This is what I understand PCMIIAW:
 ' when executed puts the execution token of the next word of the input stream onto the parameter stack.
 ['] when executed puts the execution token of the next word of the input stream onto the parameter stack.
It puts the execution token of LITERAL on the stack then compiles jsr and then the execution token of LITERAL from the parameter stack.
It leaves the execution token of the next word on the parameter stack.

```
: ['] ' [ ' literal compile, ]
; immediate
```

**Reply by jkotlinski** [reply]
`: ['] ' postpone literal ; immediate` indeed looks very pretty! I wouldn't mind a code simplification like that.

**Reply by lonetech** [reply]
@Whammo Starts out right, but `[']` doesn't leave a stack effect, it leaves compiled code by calling `literal`. That shifts the stack effect from compilation time of e.g. `irqtest` to its run time. This is why the postponement recipe compiled the `literal` call into `[']`. The only time the execution token of `literal` was on the stack was while `[']` was being compiled. So:

1. When compiling `[']`, `literal` was postponed to when `[']` runs, after the call to `'`. This counters the property that `literal` is `immediate`. 
2. `[']` was made `immediate`, so it runs during compilation of e.g. `irqtest`.
3. When compiling `irqtest`, `[']` runs `'` immediately but postpones the stack effect (its result) using `literal`.
4. When `irqtest` runs, the execution token for `flashborder` is put on the stack by the `lit` call `literal` compiled into `irqtest`.

`literal` works like: `: literal ['] lit compile, , ; immediate`


@jkotlinski Feel free to use it; it compiled to the same code when I tried it. The same recipe also occurs in `[char]` (which seems redundant, as we have character literals). 

However, you'll need to define `postpone` earlier, and that means making it not require `[']` for non-immediate words. Something like changing `] ['] compile,` to `' compile, literal ]`. 

By the way, `postpone undefined-word` somehow compiled an infinite loop. This is an ambiguous condition in the standard, so that's not a bug, though `'` does check for this error.

**Reply by Whammo** [reply]
If I'm going to get this, I suppose I'm going to have to start writing compiling words.
 I had a similar problem grasping the assembler words, but one day, I could just do it.

**Reply by lonetech** [reply]
Then you have a grasp of most of the compiling words already! In fact, I think `jsr,` and `compile,` are synonyms. The new mechanism here is immediate execution, either by `immediate` marks in the dictionary or `[` `]` state switches (`[` works like `: [ 0 state ! ; immediate`). Took me a while to understand what `postpone` was doing to non-immediate words (since postpone itself is immediate, it had to wrap one compile in another). `code` starts a definition but doesn't switch to compilation, which is why inline assembly is wrapped in `[` `]`.

**Reply by jkotlinski** [reply]
https://github.com/jkotlinski/durexforth/pull/446 looks fine?

**Reply by jkotlinski** [reply]
A note on "ambiguous condition". I thought it meant same thing as C undefined behavior, which means all bets are off. But on reading the standard, I found "ambiguous condition" has a more specific meaning.

 * "Ambiguous condition: A circumstance for which this standard does not prescribe a specific behavior."
 * A system shall document the system action taken upon each of the general or specific ambiguous conditions identified in this standard. 
 
It seems like there is some homework to do, especially on the documentation part.
(created https://github.com/jkotlinski/durexforth/issues/447)

**Comment by Whammo** [comment]
This one leverages the BRK instruction to push the stack frame.
 The IRQ routine runs first, re enables the interrupt, then your code executes.
 ~If your code takes too long it could overflow the processor stack.~

```forth
: irqpeek $dc0e c@ ;
: irqpoke $dc0e c! ;
: irqoff irqpeek $fe and irqpoke ;
: irqon irqpeek 1 or irqpoke ;

: irqsub ( addr -- )
$314 ! irqon ; \ irq vector

: brksub ( addr -- )
irqoff $316 ! ; \ brk vector

code flashborder  \ not so arbitrary asm code
$314 lda, pha, $315 lda, pha,     \ flashborder to stack
$31 lda,# $314 sta, $ea lda,# $315 sta, \ restore irq vector
      brk, nop,   \ stack for me, do irq routine
\ user code
$d020 inc,        \ return from irq routine here
$d020 dec,        \ then exit
\ end user code
sei, \ cia timer IRQs were held off until reset by irq routine
pla, $315 sta, pla, $314 sta, \ flashborder to  irq -mind the stack
$ea81 jmp,  \ restore regs, rti
;code

: irqtest
$ea31 brksub           \ call irq routine on a brk
['] flashborder irqsub \ call my code on irq
;
```

**Comment by ekipan** [comment]
Just a small thing, you can replace `irqon` `irqoff` with `1 irq!` `0 irq!`:
```forth
: irq! ( flag -- ) $dc0e c@ swap
 if 1 or else $fe and then $dc0e c! ;
```
And a pair of more general words for masking bit registers:
```forth
: bits@ ( mask addr -- bits ) c@ and ;
: bits! ( flag mask addr -- ) 
 rot >r dup c@ rot ( addr val mask ) r> 
 if or else invert and then swap c! ;
```
Edit: Whoops, `bits!` was wrong.

**Reply by Whammo** [reply]
Wow, you're really good at this stuff!
🥇

**Reply by ekipan** [reply]
You say, moments after I edit the correction :P 

I had actually written these bit masking words for my game of life program, but it was godly slow so I switched to storing whole bytes.

**Reply by Whammo** [reply]
I guess I kinda write in the idiom of the machine and brute force it. My amateur-ity shows, but 
hey, it works! 😄

**Reply by ekipan** [reply]
Your direct code has the advantage of being faster. Being more general means more faffing about `swap`ping and `rot`ting the stack, using up cycles.

**Reply by Whammo** [reply]
There is that too. It's lots of fun most of all. 
 My c64 knows who's boss.  😄
